### DISCUSSION TITLE
Add Classes

### DISCUSSION THREAD

**Discussion by lyxal** [Vyxal/Vyxal#251] [Ideas]

**Comment by ysthakur** [comment]
Once we finish basic classes, I have the following lower-priority feature requests:
- Shorter constructors, either like Scala or like C++ (I lean heavily toward the former)
- Extending/opening classes like Ruby using the same class syntax, instead of saying `Foo.bar = ...` for monkey-patching
- Letting classes define identities for addition/multiplication, overload builtins themselves, etc.

**Comment by lyxal** [comment]
I'll come up with some syntax examples for y'all and see what y'all like best

**Comment by lyxal** [comment]
```
¨@ClassName|
  →Attribute
  ←PrivateAttribute
  @Method:args|
    code
  ;
  @_Method:args
    # Private method
  ;
  
  @ClassName:args|
    # Constructor
  ;
  
  ¨∇d| # Defines an overload for `d`
    ←Attribute 4 *
  ;
  ¨∇+| # Defines an overload for `+`
    # Format: type1:type2|code
    # The stack for each overload will start as [lhs, rhs]
    
    this:int| ←Attribute +
    this:str| `No.`
    this:this| @Attribute; ←Attribute +
    otherClass:this| ¨∧ @SomeAttribute; ¨∨ @Attribute; + 
    
    # ¨∧ explicitly refers to the lhs item
    # ¨∨ explicitly refers to the rhs item
  ; 
;

¨@ClassName; # New instance of ClassName
# Methods are accessed by calling functions on that item:

¨@ClassName; @Method;

# Attributes are accessed in the same way too

¨@ClassName; @Attribute;

# Calling some operator overloads

¨@ClassName; d ,
¨@ClassName; 5 + ,
```

**Reply by lyxal** [reply]
Note that you don't need the newlines

**Reply by lyxal** [reply]
```
¨@ClassName:ParentClassA:ParentClassB...|@∷Constructor_Parameters|...;@!Inheritance_Priority_Function|...;¨→Member_Set¨←Member_Get@Method|...;¨↑Member_Equals_None;
``` 

is equivalent

**Reply by lyxal** [reply]
```
¨@A| @f|`A`; @g|`a`; ;
¨@B| @!f|`B`; @g|`b`; ;
¨@C:A:B|; # Class bodies can be empty

¨@C; ¨.f # → `B`
¨@C; ¨.g # → `a`
```

**Reply by razetime** [reply]
do we want the  entirety of Python OOP to be available in vyxal?

**Comment by chunkybanana** [comment]
I feel like using `@` instead of `¨` in most cases would be more intuitive - because function names need to be alphabetical, the spec still works fine, and `¨` is meant to be a digraph character, so this could get a bit  confusing.

**Reply by lyxal** [reply]
However I've always considered using `¨` for digraphic modifiers/transformers, as well as syntax stuff

**Reply by lyxal** [reply]
So imo, it'd be suitable

**Comment by zoomlogo** [comment]
What if we decided to interpret Vyxal instead of transpiling it? How would we handle the classes then?

**Reply by lyxal** [reply]
that'd be part of a whole different discussion, because that's part of the plan for version 3

**Comment by pxeger** [comment]
I'm not sure on classes, but if we add them, *please no inheritance*. Multiple inheritance is too complicated, and single inheritance is too restrictive. Composition is always better.

**Reply by cgccuser** [reply]
*No* inheritance is even more restrictive than single inheritance. Inheritance can often come in handy even if it's sometimes overused. And until Vyxal gets pattern matching, classes would not be super useful without inheritance. In the meantime, inheritance gives you dynamic dispatch, which lets you do cool stuff.

**Reply by pxeger** [reply]
You don't need inheritance for dynamic dispatch?

**Reply by cgccuser** [reply]
You don't, but it makes it easier

**Comment by cgccuser** [comment]
I have a small suggestion to modify Lyxal's proposal a bit to make it golfier. Instead of having a separate constructor, we could treat the entire class's body as the body of the constructor, kinda like Scala.
```
¨@ClassName:constructorParam:.attribute:|
  ←constructorParam ←attribute + →_somePrivateAttribute
  ←_somePrivateAttribute somecond [ 1 →attributeThatOnlySomeObjectsGet | 2 → attributeThatOtherObjectsGet ]

  # No separate constructor needed below
;
```
Everything else is left the same. This simply makes the class definition a little golfier. I'm thinking the class body can look almost exactly like a function body (plus overload definitions), but any variable you assign or function you create inside that class body basically becomes a field of that object. As you can see above, you don't need to statically declare them or anything, you can assign fields inside if statements and stuff so only some objects get them.

The first `constructorParam` parameter is just a normal parameter to the class's constructor, but the second, `attribute`, automatically also becomes a field of that class because of the preceding `.` (that bit of sugar isn't necessary, but it's convenient).

Also, instead of having proper private variables, we could just use underscores like Python does.

**Comment by PlaceReporter99** [comment]
Who needs classes in a golfing language? (Unless you’re doing a KOTH tournament, but you don’t need to golf it.)

**Comment by lyxal** [comment]
Done
